spackbot/handlers/style.py

```diff
@@ -23,3 +30,98 @@ async def style_comment(event, gh):
                 number,
             )
             await gh.post(comments_url, {}, data={"body": comments.style_message})
+
+
+def is_up_to_date(output):
+    """
+    A commit can fail if there are no changes!
+    """
+    return "branch is up to date" in output
+
+
+async def fix_style(event, gh):
+    """
+    Respond to a request to fix style.
+    We first retrieve metadata about the pull request. If the request comes
+    from anyone with write access to the repository, we commit, and we commit
+    under the identity of the original person that opened the PR.
+    """
+    response = requests.get(event.data["issue"]["pull_request"]["url"])
+    pr = response.json()
+
+    # Get the sender of the PR - do they have write?
+    sender = event.data["sender"]["login"]
+    repository = event.data["repository"]
+    collaborators_url = repository["collaborators_url"]
+
+    # If they don't have write, we don't allow the command
+    if not await helpers.found(gh.getitem(collaborators_url, {"collaborator": sender})):
+        logger.info(f"Not found: {sender}")
+        return (
+            "Sorry %s, I cannot do that for you. Only users with write can make this request!"
+            % sender
+        )
+
+    # Tell the user the style fix is going to take a minute or two
+    message = "Let me see if I can fix that for you! This might take a moment..."
+    await gh.post(event.data["issue"]["comments_url"], {}, data={"body": message})
+
+    # Get the username of the original committer
+    user = pr["user"]["login"]
+    email = "%s@users.noreply.github.com" % user
+
+    # We need to use the git url with ssh
+    branch = pr["head"]["ref"]
+    full_name = pr["head"]["repo"]["full_name"]
+    clone_url = "git@github.com:%s.git" % full_name
+
+    # At this point, we can clone the repository and make the change
+    with helpers.temp_dir() as cwd:
+        git("clone", "-b", branch, clone_url)
+        os.chdir("spack")
+        git("config", "user.name", user)
+        git("config", "user.email", email)
+
+        # This will authenticate the push with the added ssh credentials
+        git("remote", "set-url", "origin", clone_url)
+
+        # We need to add develop remote for this to work
+        git("remote", "add", "upstream", "https://github.com/spack/spack")
+        git("fetch", "upstream")
+
+        # This won't work if we are already on a develop branch
+        if branch != "develop":
+
+            git("checkout", "--track", "upstream/develop")
+            git("checkout", branch)
+
+        # Add newly cloned `spack` to PATH
+        os.environ["PATH"] = f"{cwd}/spack/bin:" + os.environ["PATH"]
+        from sh import spack
```

Comment:
This runs code from the PR, which is a security hole.  The PR author can change the `spack style` code in their PR to run anything they want as `spackbot`.

I think one way to do this is:

1. Clone `develop` first
2. Add the `develop` clone to `PATH`, then `from sh import spack`
3. Make a copy of the cloned repo
4. Add `clone_url` (which should probably be called `fork_url` or something since we're not cloning it now) as a remote in the copy
5. Check out `branch` in the copy (the one that is *not* in `PATH`) to avoid a second full clone (which is slow)
6. Run `spack style --root .` in the clone

I'm adding a `--root` option to `spack style` so that we can do this.  It will make the following changes:

* Run in the clone's prefix
* Be sure to use the invoked `spack` instance's configs for the various tools (i.e., don't allow `flake8`, `isort`, `black`, `mypy`, etc. to use configs from the clone (`flake8`'s config in particular can specify additional python code files to run)

We should probably also have spackbot decline to format PRs that modify style configs or the `spack style` command, as they won't work if we're only running `spack style` from `develop` -- users making changes to those would need to run `spack style` themselves.

Having written all that, the other option that I kind of like a lot better would be to have `spackbot` run `spack style` in a container as an unprivileged user.  Do you know how to do that @vsoch?  We could bindmount the clone and run its `spack style` in an environment that does not have access to the `spackbot`'s credentials. If we could do that we could avoid some of this complexity.  Basically I do not want the code from the PR to be able to get at private keys, tokens, `id_spackbot`, etc.

@adamjstewart @becker33 @alalazo thoughts on this?